flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/LookupJoinUtil.java

```diff
@@ -144,75 +146,92 @@ private LookupJoinUtil() {
             //  currently we only support top-level lookup keys
             int[][] indices =
                     IntStream.of(lookupKeyIndicesInOrder)
                             .mapToObj(i -> new int[] {i})
                             .toArray(int[][]::new);
+
             LookupTableSource tableSource =
                     (LookupTableSource) ((TableSourceTable) temporalTable).tableSource();
             LookupRuntimeProviderContext providerContext =
                     new LookupRuntimeProviderContext(indices);
             LookupTableSource.LookupRuntimeProvider provider =
                     tableSource.getLookupRuntimeProvider(providerContext);
 
-            if (requireSyncLookup && !(provider instanceof TableFunctionProvider)) {
-                throw new TableException(
-                        String.format(
-                                "Require a synchronous TableFunction due to planner's requirement but no TableFunctionProvider "
-                                        + "found in TableSourceTable: %s, please check the code to ensure a proper TableFunctionProvider is specified.",
-                                temporalTable.getQualifiedName()));
-            }
             if (provider instanceof LookupFunctionProvider) {
                 if (provider instanceof PartialCachingLookupProvider) {
                     PartialCachingLookupProvider partialCachingLookupProvider =
                             (PartialCachingLookupProvider) provider;
-                    return new CachingLookupFunction(
-                            partialCachingLookupProvider.getCache(),
-                            partialCachingLookupProvider.createLookupFunction());
+                    syncLookupFunctions =
+                            new CachingLookupFunction(
+                                    partialCachingLookupProvider.getCache(),
+                                    partialCachingLookupProvider.createLookupFunction());
+                } else {
+                    syncLookupFunctions =
+                            ((LookupFunctionProvider) provider).createLookupFunction();
                 }
-                return ((LookupFunctionProvider) provider).createLookupFunction();
-            } else if (provider instanceof AsyncLookupFunctionProvider) {
+            }
+            if (provider instanceof AsyncLookupFunctionProvider) {
                 if (provider instanceof PartialCachingAsyncLookupProvider) {
                     PartialCachingAsyncLookupProvider partialCachingLookupProvider =
                             (PartialCachingAsyncLookupProvider) provider;
-                    return new CachingAsyncLookupFunction(
-                            partialCachingLookupProvider.getCache(),
-                            partialCachingLookupProvider.createAsyncLookupFunction());
+                    asyncLookupFunctions =
+                            new CachingAsyncLookupFunction(
```

Comment:
this function should be refactored to create nested delegators of user lookup func

[CachingLookupFunction]
                   |
[RetryableLookupFunctionDelegator]
                   |
         userLookupFunc

I'll update this